### PR TITLE
Issue/5380 background refresh on reader

### PR DESCRIPTION
**Pull request by elibud**
**Fixes** #5380 

**To test:**

**A) Test on the Simulator**

Regular Background Fetch
1. Start the app from XCode
2. Go to Reader->AnyFeed (I suggest a8c feed since there's always sth happening there)
3. Simulate Home button
4. XCode->Debug->Simulate Background Fetch

App Started from a Background Fetch
1. Edit Run Scheme, select Background Fetch: Launch App due to a background fetch event

On both cases you can check that the refresh was performed by looking at the console, and of course if new content was fetched it should be present on the reader.

**B) Test on the Device**

You can perform the aforementioned tests on the device, but is not too much fun. What is great is when you go into the app and the content is updated :)

1. Deploy the app to your device and go to one of your reader feeds.
2. Press home and continue with your life.
3. After a few hours go back to the app and check the feed, it should be beautifully updated.

Info on background refresh can be found on https://developer.apple.com/videos/play/wwdc2013/204/ minute 8 to 22

Needs review: @aerych

### CODE REVIEW

**Comment by elibud** [comment]
I left the logs for debugging only, I'll remove it when we are happy with the changes and functionality.

**Comment by elibud** [comment]
> My understanding is that during a background fetch if there is new data the app will attempt to take an > updated snapshot of the UI to display during a subsequent launch. Did I understand that correctly?
Yes that is correct. The idea is that the snapshot is used on multitasking so the user can see that they have new content.

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 1]
```diff
@@ -1086,6 +1091,30 @@ import WordPressComAnalytics
         }
     }
 
+    func backgroundFetch(_ completionHandler: ((UIBackgroundFetchResult) -> Void)?) {
+        print("=========>>>>>>>>>>> backgroundFetch")
+        print("=========>>>>>>>>>>> \(String(describing: readerTopic?.title))")
+
+        let lastSeenPostID = (tableViewHandler.resultsController.fetchedObjects?.first as? ReaderPost)?.postID
```
Comment (elibud) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: How would you suggest doing that? I want to continue even if I didn't have any content loaded before, right? Even though I can not think of any reasonable circumstance for that to happen...

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 2]
```diff
@@ -1086,6 +1091,30 @@ import WordPressComAnalytics
...
+
+        syncHelper.backgroundSync(success: { [weak self, weak lastSeenPostID] in
+            let newestFetchedPostID = (self?.tableViewHandler.resultsController.fetchedObjects?.first as? ReaderPost)?.postID
```
Comment (elibud) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: I always have to execute the completionHandler so exiting early is not a good idea here.

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 3]
```diff
@@ -1086,6 +1091,30 @@ import WordPressComAnalytics
         }
     }
 
+    func backgroundFetch(_ completionHandler: ((UIBackgroundFetchResult) -> Void)?) {
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: Does the completion handler need to be an optional?

Comment (elibud) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: I was following what we were doing on PushNotificationsManger, but in this case the correct thing to do is to have it as non-optional and @escaping, will update accordingly. Thx

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift (hunk 2 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: My concern here, and on ln 1098 above, is the optional chain allows for something declared with a `let` to be set to `nil`, which doesn't seem like a good practice.  Maybe I'm being too picky?

Comment (elibud) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: This is exactly why I don't really like optionals. Do you think it is better if we give it a default value? Something like: 


```        
let lastSeenPostID = (tableViewHandler.resultsController.fetchedObjects?.first as? ReaderPost)?.postID ?? -1
syncHelper.backgroundSync(success: { [weak self, weak lastSeenPostID] in
    let newestFetchedPostID = (self?.tableViewHandler.resultsController.fetchedObjects?.first as? ReaderPost)?.postID ?? -1
    self?.tableView.scrollToRow(at: IndexPath(row: 0, section: 0), at: .top, animated: false)
    if lastSeenPostID == newestFetchedPostID {
```

Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: That works for me. :)

**Comment by elibud** [comment]
@aerych All the requested changes done and logs removed. Tested on Alpha and Internal, ready for the final pass.

**Comment by aerych** [comment]
Nice patch :)
Just a final nitpick, it might be nice to have documentation for the new `backgroundFetch` method. Otherwise this looks great. Marking approved and :shipit: when ready!

**Pull request merged**